### PR TITLE
[AIRFLOW-4268] Add MsSqlToGoogleCloudStorageOperator

### PR DESCRIPTION
**Pull request by Tomme**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [AIRFLOW-4268](https://issues.apache.org/jira/browse/AIRFLOW-4268) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4268

### Description

- [x] Added MsSqlToGoogleCloudStorageOperator using MsSqlHook and GoogleCloudStorageHook.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
  - tests/contrib/operators/test_mssql_to_gcs_operator.py

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by OmerJog** [comment]
My 2 notes:
1. I think it's better to support all the relevant argument of the hook. For example the operator introduced in this PR doesn't support the gzip flag of the  GoogleCloudStorageHook (reference https://github.com/apache/airflow/pull/3893 ) 

2. <strike> PR https://github.com/apache/airflow/pull/5043 raised a question about the num_retries of the hook/operator. I didn't see the committers agree on course of action hopefully it will be done soon and this PR will be updated accordingly</strike>.  No longer relevant after https://github.com/apache/airflow/pull/5054

File: tests/contrib/operators/test_mssql_to_gcs_operator.py
Comment (XD-DENG) [inline tests/contrib/operators/test_mssql_to_gcs_operator.py:27]: In master branch we already dropped support to Python 2. So no need to have this (and the related lines)

**Review by XD-DENG — CHANGES_REQUESTED** [review]

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (XD-DENG) [inline airflow/contrib/operators/mssql_to_gcs.py]: type of `delegate_to` is missing here.

**Review by XD-DENG — CHANGES_REQUESTED** [review]

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py]: Resolved with [797c4e1](https://github.com/apache/airflow/pull/5077/commits/797c4e11c0c4eab873ed8e3a4843646665d6b407).

File: tests/contrib/operators/test_mssql_to_gcs_operator.py
Comment (Tomme) [reply tests/contrib/operators/test_mssql_to_gcs_operator.py:27]: Resolved with [797c4e1](https://github.com/apache/airflow/pull/5077/commits/797c4e11c0c4eab873ed8e3a4843646665d6b407).

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (kaxil) [inline airflow/contrib/operators/mssql_to_gcs.py]: Need to move docstrings to Line 40

**Review by kaxil — COMMENTED** [review]
Minor change suggest otherwise LGTM. Good work @Tomme

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py]: Resolved with [a9cdc5a](https://github.com/apache/airflow/pull/5077/commits/a9cdc5ad1bd8d60c910c80c6a9cdea4fa87e7302).

**Comment by Tomme** [comment]
Cheers @kaxil, all squashed now 👍

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (kaxil) [inline airflow/contrib/operators/mssql_to_gcs.py]: Let's not use `object` as it is a built-in name.. Change it to `object_name` for example or `file`

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (kaxil) [inline airflow/contrib/operators/mssql_to_gcs.py:78]: May be add an example of using this operator in docstrings... Similar to https://github.com/apache/airflow/blob/master/airflow/contrib/operators/gcs_list_operator.py#L51

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (kaxil) [inline airflow/contrib/operators/mssql_to_gcs.py:131]: Missing blank space

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (kaxil) [inline airflow/contrib/operators/mssql_to_gcs.py:40]: Add example of this case as well on how to use `{}` in docstrings

**Review by kaxil — CHANGES_REQUESTED** [review]

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (RosterIn) [inline airflow/contrib/operators/mssql_to_gcs.py]: This is wrong. The limit is not 4GB it's 5GB:
https://cloud.google.com/storage/quotas
"There is a maximum size limit of 5 TB for individual objects stored in Cloud Storage."
I think in general It's best not to mention it at all. The limits can change at will. Airflow can't keep track of all size changes.

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (RosterIn) [inline airflow/contrib/operators/mssql_to_gcs.py]: What conversion is done in this function? It gets `value `param  and return the same `value`.

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (RosterIn) [inline airflow/contrib/operators/mssql_to_gcs.py:233]: Can you please add support for some more very used types?
For example MsSQL has:  money ,smallmoney, smalldatetime  etc...
These types needs to be mapped to the best match in BigQuery types.

**Review by RosterIn — CHANGES_REQUESTED** [review]

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py:233]: As per: [_mssql.pyx#L1710](https://github.com/pymssql/pymssql/blob/fda5fd38f30eaeeb40a911235692144493d69864/src/_mssql.pyx#L1710) `pymssql` only returns five basic DB types.

Comment (RosterIn) [reply airflow/contrib/operators/mssql_to_gcs.py:233]: yep. apparently the package handle these transforms on it's own. It converts money to DECIMAL.

Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py:233]: We could use `_mssql` over `pymssql` but it is not DB-API compliant plus the `MsSqlHook` would need changing too.

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py]: Ah it was made redundant when I implemented `class DecimalEncoder(json.JSONEncoder)`, however I have now removed `DecimalEncoder` and reimplemented `convert_types`, keeping the design of this operator closer to other similar ones.

Resolved with [97c9cd7](https://github.com/apache/airflow/pull/5077/commits/97c9cd711f958f27dd82fb00f9fab334892f3ff1).

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py]: Agreed, bad wording from old BigQuery limitations.

Resolved with [97c9cd7](https://github.com/apache/airflow/pull/5077/commits/97c9cd711f958f27dd82fb00f9fab334892f3ff1).

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py:40]: Resolved with [97c9cd7](https://github.com/apache/airflow/pull/5077/commits/97c9cd711f958f27dd82fb00f9fab334892f3ff1).

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py:131]: Resolved with [97c9cd7](https://github.com/apache/airflow/pull/5077/commits/97c9cd711f958f27dd82fb00f9fab334892f3ff1).

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py:78]: Resolved with [97c9cd7](https://github.com/apache/airflow/pull/5077/commits/97c9cd711f958f27dd82fb00f9fab334892f3ff1).

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [reply airflow/contrib/operators/mssql_to_gcs.py]: Resolved with [97c9cd7](https://github.com/apache/airflow/pull/5077/commits/97c9cd711f958f27dd82fb00f9fab334892f3ff1).

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (RosterIn) [reply airflow/contrib/operators/mssql_to_gcs.py:233]: Maybe better leave it to the users. The operator accept SQL query. The user can convert the type of the  "non supported" columns to something else within the query (like TEXT or whatever).
For example:
https://stackoverflow.com/questions/532351/how-do-i-convert-from-a-money-datatype-in-sql-server

**Comment by Tomme** [comment]
@OmerJog Regarding your first point (and PR #3893), I agree it would be good to support gzip however what would be the best approach to take? Should we improve operators on an ad-hoc basis or in bulk?

**Comment by OmerJog** [comment]
@Tomme This is a new operator to the repo so I think it's better to have it with as much functionality as possible. However there is also the question of your time.

File: airflow/contrib/operators/mssql_to_gcs.py
Comment (Tomme) [inline airflow/contrib/operators/mssql_to_gcs.py:211]: @OmerJog Agree, I've implemented a quick one-liner simply to avoid rebuilding to much, but imagine should be fine? (BigQuery will not like compressed schema files)

Comment (OmerJog) [reply airflow/contrib/operators/mssql_to_gcs.py:211]: Yep. There is no need to compress the schema file.
Compression is for the data files.

**Comment by OmerJog** [comment]
@kaxil PTAL

**Comment by kaxil** [comment]
Thanks @OmerJog

**Comment by stpavinash** [comment]
I'm trying to connect to sql server instance using this operator and trying to a query a table which has got 20M records and the airflow task is in running state for atleast 8 hours and it never generated a json file in GCS. However, tables that have like 200k records were queried in like 1 min and json was created in GCS. Could you please help?

**Comment by mik-laj** [comment]
Can you create a bug report in Jira? Please add the "gcp" component.

**Pull request merged**
